cpp/src/arrow/builder.h

```diff
@@ -847,6 +847,8 @@ Status ARROW_EXPORT MakeDictionaryBuilder(MemoryPool* pool,
                                           const std::shared_ptr<DataType>& type,
                                           std::shared_ptr<ArrayBuilder>* out);
 
+Status ARROW_EXPORT MakeDictionaryArray(MemoryPool* pool, const Array& input,
+                                        std::shared_ptr<Array>* out);
```

Comment:
Can you flip the argument order here

`(const Array& input, MemoryPool* pool, ...)`

(Google C++ style recommends argument ordering of: in args, in-out args, out-args)

How about a more descriptive name like `EncodeDictionary` (or `EncodeToDictionaryArray`)? Can you also add a doxygen comment with at least a `\brief` description